### PR TITLE
Implement Smart Router API

### PR DESCRIPTION
**Pull request by odcheung**
The Whirlpools Smart Router API aims to help developers find, select and execute the best trade route between 2 tokens through a set of Whirlpools. 

The router allows the user to:
- swap on routes with up to 2-hops
- swap on routes that is split into multiple smaller routes
    - A full trade can be broken into multiple small trades for better price performance
- allow user to choose the best executable route based on the current execution environment. Variables include
    - Wallet supported transaction version
    - Whether the wallet needs to initialize ATAs (or wrap SOL) to perform the swap
    - Availability of lookup tables for the route's touched accounts 

### Building a Whirlpool Router
There are 3 options to build a `WhirlpoolRouter`
```typescript
const poolAddresses: Address[] = [];

// 1. Building WhirlpoolRouter with a set of Whirlpool Addresses
const router = await WhirlpoolRouterBuilder.buildWithPools(ctx, poolAddresses);

// 2. Building WhirlpoolRouter with a pre-generated poolgraph
const graph = await PoolGraphBuilder.buildPoolGraphWithFetch(...poolAddresses);
const router = WhirlpoolRouterBuilder.buildWithPoolGraph(ctx, graph);

// 3. Get a router instance from a WhirlpoolClient
const whirlpoolClient: WhirlpoolClient;
const router = await whirlpoolClient.getRouter(poolAddresses)
```

### [Optional]Set up a LookupTableFetcher

The new router supports atomic trades up to 2 pools in a single transaction. Given the amount of extra data, we need to utilize [Solana Address Lookup tables](https://docs.solana.com/developing/lookup-tables) to minimize the Transaction size. 

The new system takes in an optional LookupTableFetcher. It will use this fetcher to fetch pre-indexed `AddressLookupTableAccount` for the addresses involved in the route transaction. If a LookupTableFetcher is provided and it successfully returns indexed LookupTable accounts, the more executable options will be available to you during the route-selection process.

We provide an out of the box lookupFetcher that indexes lookup tables for Orca's Whirlpools - `OrcaLookupTableFetcher` in `@orca-so/orca-sdk - v0.1.0`.
```typescript
const lookupTableFetcher = new OrcaLookupTableFetcher(axios, connection);
const ctx = WhirlpoolContext.from(..., lookupTableFetcher)
// Build you WhirlpoolClient or WhirlpoolRouter with this ctx
// ...
// When you execute the best-route transaction, pass on the lookupTable accounts into the build method call.
const [bestRoute, lookupTablesAccounts] = bestRouteResult;
const swapTxBuilder = await router.swap(bestRoute, slippageTolerance, walletATA);
const txId = await swapTxBuilder.buildAndExecute({
    maxSupportedTransactionVersion: 0,
    lookupTablesAccounts
})
```

## Finding, selecting and executing Routes

The process from defining a trade to executing it can be broken down into a few steps
1. Generate all possible routes from a pool-graph
2. Selecting the best route that is executable under the current execution environment. Most routes are limited by variables that affects the total Transaction byte-size.
3. Once the best route is selected, generating the swap transaction to execute.

### Use-case 1 - Immediate execution of routes
Suitable for use-cases where routes selection happens immediately and without user-delay (ex. scripts)

```typescript
const walletATA: AccountInfo[] = [] // Populate with ATA accounts from wallet

// Construct a trade from 1000 USDC to SOL
const trade: Trade = {
	tokenIn: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v", // USDC
	tokenOut: "So11111111111111111111111111111111111111112", // SOL
	tradeAmount: new u64("1000000000", // 1000 USDC
	amountSpecifiedIsInput: true
}
const slippageTolerance = Percentage.fromFraction(1, 100); // 1% slippage

const bestRouteResult = await router.findBestRoute(trade, true, undefined, {
	maxSupportedTransactionVersion: 0,
	availableAtaAccounts: walletATA
});
// ...null check bestRouteResult
const [bestRoute, lookupTablesAccounts] = bestRouteResult;
const swapTxBuilder = await router.swap(bestRoute, slippageTolerance, walletATA);

let buildOptions: Partial<BuildOptions> = {};
if (wallet.supportsV0()) {
	buildOptions = {
		maxSupportedTransactionVersion: 0,
		lookupTablesAccounts
	};
} else {
	buildOptions = {
		maxSupportedTransactionVersion: "legacy",
	}
}
const txId = await swapTxBuilder.buildAndExecute(buildOptions)

```

## Use-case 2 - Deferred execution of route
For use cases where there's a deferred execution of the route after selection.
Most common use case include websites presenting the best route / quote before the user chooses to execute.

The difference here is that the regular `router.findBestRoute` call is broken into `router.findAllRoutes` and `RouterUtils.selectFirstExecutableRoute`. The developer has to call the `select` call every time prior to execution as the execution environment may have changed since the last evaluation.

```typescript
const lookupTableFetcher = new OrcaLookupTableFetcher(axios, connection);
const ctx = WhirlpoolContext.from(..., lookupTableFetcher)
const graph = PoolGraphBuilder.buildPoolGraph();
const router = WhirlpoolRouterBuilder.buildWithPoolGraph(ctx, graph);
const walletATA: AccountInfo[] = [] // Populate with ATA accounts from wallet


// Construct a trade from 1000 USDC to SOL
const trade: Trade = {
	tokenIn: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v", // USDC
	tokenOut: "So11111111111111111111111111111111111111112", // SOL
	tradeAmount: new u64("1000000000", // 1000 USDC
	amountSpecifiedIsInput: true
}
const slippageTolerance = Percentage.fromFraction(1, 100); // 1% slippage

// [Route Presentation] Selecting the best route to present
// The reason we are not using findBestRoute is because the execution environment may change after quote presentation. (ex. wallet change). Best practice is to re-select the best route from this generated route set during execution time.
const routes = await router.findAllRoutes(trade, true);
const bestRouteResult = await RouterUtils.selectFirstExecutableRoute(ctx, routes, {
	// "legacy if wallet does not support v0 tx"
	maxSupportedTransactionVersion: 0,
	availableAtaAccounts: walletATA
}
// ...
const [bestRoute, lookupTablesAccounts] = bestRouteResult;

// ...Present bestRoute to user here

// [Route Execution] Some time has passed. User has decided to execute on route
// NOTE: WhirlpoolContext may have changed (ex. wallet change). If so, regenerate anything related to the changes. Such as ctx, router, walletATA.
const [bestRoute, lookupTablesAccounts] = await RouterUtils.selectFirstExecutableRoute(ctx, routes, {
	// "legacy if wallet does not support v0 tx"
	maxSupportedTransactionVersion: 0,
	availableAtaAccounts: walletATA
}

const swapTxBuilder = await router.swap(bestRoute, slippageTolerance, walletATA);

let buildOptions: Partial<BuildOptions> = {};
if (wallet.supportsV0()) {
	buildOptions = {
		maxSupportedTransactionVersion: 0,
		lookupTablesAccounts
	};
} else {
	buildOptions = {
		maxSupportedTransactionVersion: "legacy",
	}
}
const txId = await swapTxBuilder.buildAndExecute(buildOptions)
```

## To be improved:
- Route evaluation does not take into account of transaction cost (cost of initializing ATAs, cost to execute etc)

Smart-router internal logic design & author - @philcchen

### CODE REVIEW

File: sdk/src/router/public/router-utils.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+      const tx = await getSwapFromRoute(
+        ctx,
+        {
+          route,
+          slippage: opts.slippageTolerance,
+          resolvedAtaAccounts: opts.availableAtaAccounts ?? null,
+          wallet: wallet.publicKey,
+        },
+        false
+      );
```
Comment (yugure-orca) [inline sdk/src/router/public/router-utils.ts:79]: To reserve tx space for priority fee related instructions...
```
if (opts.instructionsToReserveTxSize) {
    tx.addInstruction(opts.instructionToReserveTxSize);
}
```

File: sdk/src/errors/errors.ts [hunk 2]
```diff
@@ -22,15 +22,27 @@ export enum SwapErrorCode {
   TickArraySequenceInvalid = `TickArraySequenceInvalid`,
 }
 
-export type WhirlpoolsErrorCode = TokenErrorCode | SwapErrorCode | MathErrorCode;
+export enum RouteQueryErrorCode {
```
[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/src/errors/errors.ts:25]: yep! makes sense

File: sdk/src/router/quote-map.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+  programId: PublicKey,
+  fetcher: AccountFetcher,
+  opts: RoutingOptions
+) {
+  const { percentIncrement, numTopPartialQuotes } = opts;
+  const { tokenIn, tokenOut, tradeAmount } = trade;
+
+  const { percents, amounts } = getSplitPercentageAmts(tradeAmount, percentIncrement);
+  // The max route length is the number of iterations of quoting that we need to do
+  const maxRouteLength = Math.max(...paths.map((path) => path.edges.length));
```
[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/src/router/quote-map.ts:39]: good catch!

File: sdk/src/router/public/router-utils.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+ * @category Router
+ * @param slippageTolerance The slippage tolerance to use when selecting the best route.
+ * @param maxSupportedTransactionVersion The maximum transaction version that the wallet supports.
+ * @param availableAtaAccounts A list of ATA accounts that are available in this wallet to use for the swap.
+ */
+export type RouteSelectOptions = {
+  slippageTolerance: Percentage;
+  maxSupportedTransactionVersion: "legacy" | number;
+  availableAtaAccounts?: AtaAccountInfo[];
+};
```
[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/src/router/public/router-utils.ts:28]: good call. i'm actually going to expand this idea and add a callback fn here to give users a chance to manipulate the transaction builder after we are done with it, but prior to us evaluating the txnSize.

File: sdk/src/router/public/router-utils.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+          addressesToLookup
+        );
+        try {
+          v0TxSize = await tx.txnSize({
+            latestBlockhash: MEASUREMENT_BLOCKHASH,
+            maxSupportedTransactionVersion: opts.maxSupportedTransactionVersion,
+            lookupTableAccounts,
+          });
+
+          if (v0TxSize !== undefined && v0TxSize < ENCODED_LIMIT) {
```
Comment (yugure-orca) [inline sdk/src/router/public/router-utils.ts:123]: I think we need to use 1 constant for size comparison.

Now ``TransactionBuilder.txnSize`` returns the size of Base64 encoded data, so we need to use ``ENCODED_LIMIT`` for both legacy and v0.

But in my opinion, it is better to update ``TransactionBuilder.txnSize`` to return the size of UNencoded data, and use ``TX_LIMIT`` (or ``PACKET_DATA_SIZE`` from web3.js) for both legacy and v0.

**Review by yugure-orca — COMMENTED** [review]
Thanks for the update!
Almost all looks good to me.

last one: RouteUtils.selectFirstExecutableRoute (TransactionBuilder.txnSize related)
I've posted the detail in Discord 🙏

File: sdk/src/router/public/router-utils.ts (hunk 5 above)
Comment (yugure-orca) [reply sdk/src/router/public/router-utils.ts:123]: PR https://github.com/orca-so/whirlpools/pull/100 will fix this. 🙂

**Review by yugure-orca — APPROVED** [review]
Now, I believe we're ready to sail.

**Pull request merged**